### PR TITLE
[FEATURE] Reduced size of EPR dictionary

### PR DESCRIPTION
**Pull request by cpockrandt**
- avoided multi-line constexpr
- remove blocks/superblocks/ultrablocks for largest character for another if-statement
- removed the carry bit
- some code-cleanup
- some additional minor changes

### CODE REVIEW

File: include/seqan/index/index_fm_lf_table.h [hunk 1]
```diff
@@ -528,6 +553,15 @@ inline void _setSentinelSubstitute(LF<TText, TSpec, TConfig> & lf)
     lf.sentinelSubstitute = ordVal;
 }
 
+template <typename TText, typename TSpec, typename TConfig>
+inline std::enable_if_t<isLevelsPrefixRD<typename TConfig::Bwt>::value, void>
+_setSentinelSubstitute(LF<TText, TSpec, TConfig> & lf)
+{
+    // TODO(cpockrandt): check whether it is a problem
```
Comment (h-2) [inline include/seqan/index/index_fm_lf_table.h]: ?

Comment (h-2) [inline include/seqan/index/index_fm_lf_table.h]: Please do :)

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 2]
```diff
@@ -345,35 +384,50 @@ struct RankDictionaryEntry_<TValue, Levels<TSpec, TConfig> >
...
-    return (blocks == constBlocks) ?
-           _bitmask2<TWord>(bitsTotal, blocks - 1, constBlocks, blocksize, value << (bitsTotal - blocksize)) :
-           (
-               (blocks == 0) ?
-               value >> (bitsTotal % blocksize) :
-               _bitmask2<TWord>(bitsTotal, blocks - 1, constBlocks, blocksize, value | (value >> blocksize))
-           );
+    return (blocksLeft == 0) ?
+           result :
+           _bitmaskRecursive<TWord>(bitsUsed, blocksLeft - 1, blocksProcessed + 1, blocksize, value1, value2, result | (((blocksProcessed % 2) ? value1 : value2) << (bitsUsed - blocksProcessed * blocksize)));
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: I usually use this syntax for the ternary operator:
```c++
return (a < b)
    ? true
    : false;
```
Also please make sure that total line length doesn't exceed 120characters; move the return type into an extra line and put each parameter in its own line.

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 3]
```diff
@@ -399,48 +453,31 @@ constexpr TWord _bitmaskWrapper(RankDictionary<bool, Levels<TSpec, LevelsPrefixR
...
+    static constexpr const unsigned _BITS_PER_WORD    = Min<RankDictionaryWordSize_<TValue, Levels<TSpec, TConfig> >::VALUE, _BITS_PER_BLOCK>::VALUE;
+    static constexpr const unsigned _VALUES_PER_WORD  = _BITS_PER_WORD / _BITS_PER_VALUE;
+    static constexpr const unsigned _WORDS_PER_BLOCK  = _BITS_PER_BLOCK / _BITS_PER_WORD;
+    static constexpr const unsigned _VALUES_PER_BLOCK = _VALUES_PER_WORD * _WORDS_PER_BLOCK;
+    static constexpr const uint64_t _VALUES_PER_SUPERBLOCK = Max<1, (((1ull << (BitsPerValue<typename RankDictionaryBlockType_<typename TConfig::Size, TConfig::LEVELS, 2>::Type>::VALUE/2)) - 1) / _VALUES_PER_BLOCK) * _VALUES_PER_BLOCK>::VALUE; // 2^x - 1 values
+    static constexpr const uint64_t _VALUES_PER_ULTRABLOCK = Max<1, (((1ull << (BitsPerValue<typename RankDictionaryBlockType_<typename TConfig::Size, TConfig::LEVELS, 3>::Type>::VALUE/2)) - 1) / Max<_VALUES_PER_SUPERBLOCK, 1>::VALUE) * _VALUES_PER_SUPERBLOCK>::VALUE; // MAX: workaround for intel & clang: division is not constexpr since divisor could be 0
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: for values `constexpr` implies `const` so you can remove the latter.

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 4]
```diff
@@ -456,36 +493,75 @@ struct RankDictionary<TValue, Levels<TSpec, TConfig> >
     // Constructors
     // ------------------------------------------------------------------------
 
+    void _populateBitmasks()
+    {
+        // TODO(cpockrandt): make all bitmasks a const-expr
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: please do :smile:

Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: and I am not sure if they should be static...

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 5]
```diff
@@ -456,36 +493,75 @@ struct RankDictionary<TValue, Levels<TSpec, TConfig> >
...
+
+        static const unsigned maxValue = (1 << _BITS_PER_VALUE) - 1;
+        static const unsigned padding = _BITS_PER_WORD % _BITS_PER_VALUE;
+        static const TWordType paddingWord = (padding > 0) ? (1ull << (_BITS_PER_WORD - padding)) : 0ull;
+        if (isPRDict(*this) && !std::is_same<TValue, bool>::value)
+        {
+            if (padding > 0)
+                _SELECT_BITMASK = _bitmask<TWordType>(_BITS_PER_WORD, _VALUES_PER_WORD, _BITS_PER_VALUE, maxValue, 0u);
+            else
+                _SELECT_BITMASK = _bitmask<TWordType>(_BITS_PER_WORD, _VALUES_PER_WORD, _BITS_PER_VALUE, 0u, maxValue);
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: introduce newline

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 6]
```diff
@@ -456,36 +493,75 @@ struct RankDictionary<TValue, Levels<TSpec, TConfig> >
...
         createRankDictionary(*this, text);
     }
 };
 
+/*template <typename TValue, typename TSpec, typename TConfig>
+struct RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSize, TFibre, LEVELS, WPB> > >:
+    RankDictionary<TValue, Levels<TSpec, TConfig> >
+{
+
+};*/
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: ?

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 7]
```diff
@@ -959,64 +1035,77 @@ _getUltraBlockRank(RankDictionary<bool, Levels<TSpec, LevelsPrefixRDConfig<TSize
...
+    typedef RankDictionary<TValue, Levels<TSpec, TConfig> >                TRD;
 
-    TWord mask = word ^ TRankDictionary::_CHAR_BITMASKS[ordValue(c)];
+    TWord mask = word ^ TRD::_CHAR_BITMASKS[ordValue(c)];
     // NOTE: actually it should be: mask & (mask >> 1) & (mask >> 2) & ... but this is shorter and equivalent
-    for (TWord i = 1; i < TRankDictionary::_BITS_PER_VALUE; ++i)
+    for (TWord i = 1; i < TRD::_BITS_PER_VALUE; ++i)
         mask &= mask >> 1;
-    return popCount(TRankDictionary::_TRUNC_BITMASKS[posInWord] & mask);
+    return popCount(TRD::_TRUNC_BITMASKS[posInWord] & mask);
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: Why this renaming?

File: tests/index/test_index_fm_rank_dictionary.cpp [hunk 8]
```diff
@@ -78,22 +78,22 @@ typedef Levels<void, LevelsRDConfig<uint32_t, Alloc<>, 3, 3> >       Default3Lev
 typedef
     TagList<RankDictionary<bool,            Prefix1Level>,
     TagList<RankDictionary<Dna,             Prefix1Level>,
-    TagList<RankDictionary<Dna5Q,           Prefix1Level>,
+    TagList<RankDictionary<Dna5,            Prefix1Level>,
```
Comment (h-2) [inline tests/index/test_index_fm_rank_dictionary.cpp]: ?

File: include/seqan/basic/tuple_bit_compressed.h [hunk 9]
```diff
@@ -88,15 +88,13 @@ template <> struct BitVector_<255>;
...
 {
-    static const unsigned BITS_PER_VALUE = BitsPerValue<TValue>::VALUE + (std::is_same<TSpec, PlusOne>::value ? 1 : 0);
-
-    typedef typename BitVector_<SIZE * BITS_PER_VALUE>::Type TBitVector;
+    typedef typename BitVector_<SIZE * BitsPerValue<TValue>::VALUE>::Type TBitVector;
 
-    static const uint64_t BIT_MASK = ((1ull << (BITS_PER_VALUE - 1)       ) - 1ull) << 1 | 1ull;
-    static const uint64_t MASK     = ((1ull << (SIZE * BITS_PER_VALUE - 1)) - 1ull) << 1 | 1ull;
+    static const uint64_t BIT_MASK = ((1ull << (BitsPerValue<TValue>::VALUE - 1)       ) - 1ull) << 1 | 1ull;
+    static const uint64_t MASK     = ((1ull << (SIZE * BitsPerValue<TValue>::VALUE - 1)) - 1ull) << 1 | 1ull;
```
Comment (h-2) [inline include/seqan/basic/tuple_bit_compressed.h]: could be constexpr as well

File: tests/index/test_index_fm_rank_dictionary.cpp (hunk 8 above)
Comment (h-2) [inline tests/index/test_index_fm_rank_dictionary.cpp]: please reintroduce Dna5Q / add Dna5

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 10]
```diff
@@ -959,64 +1131,105 @@ _getUltraBlockRank(RankDictionary<bool, Levels<TSpec, LevelsPrefixRDConfig<TSize
...
+             TPosInWord const posInWord,
+             TValue const c)
 {
-    typedef RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSize, TFibre, LEVELS, WPB> > >    TRankDictionary;
+    typedef RankDictionary<TValue, TPrefixLevels>    TRD;
+
+    TWord erg1 = (TRD::_CHAR_BITMASKS[ordValue(c)] - (word & TRD::_SELECT_BITMASK)) & TRD::_TRUNC_BITMASKS[(posInWord + 1)/2];
+    TWord erg2 = (TRD::_CHAR_BITMASKS[ordValue(c)] - ((word >> TRD::_BITS_PER_VALUE) & TRD::_SELECT_BITMASK)) & TRD::_TRUNC_BITMASKS[(posInWord/2) + 1];
 
-    return popCount((TRankDictionary::_CHAR_BITMASKS[ordValue(c)] - word) & TRankDictionary::_TRUNC_BITMASKS[posInWord]);
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: why was the computation moved to lcal variables?

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 11]
```diff
@@ -1221,25 +1457,36 @@ getRank(RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSize, TFibre,
...
-         + _getSuperBlockRank(dict, superBlock, pos, static_cast<TValue>(c))
-         + _getBlockRank(dict, entry.block, pos, static_cast<TValue>(c))
-         + _getValueRank(dict, entry.values, posInBlock, static_cast<TValue>(c));
+    else if (ordValue(c) == 0)
+        return _getUltraBlockRank(dict, ultraBlock, pos, static_cast<TValue>(c))
+             + _getSuperBlockRank(dict, superBlock, pos, static_cast<TValue>(c))
+             + _getBlockRank(dict, entry.block, pos, static_cast<TValue>(c))
+             + _getValueRank(dict, entry.values, posInBlock, static_cast<TValue>(c));
+    else
+    {
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: please always balance braces, i.e. add them to the above codeblocks, even if they contain only one statement.

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 12]
```diff
@@ -399,48 +473,38 @@ constexpr TWord _bitmaskWrapper(RankDictionary<bool, Levels<TSpec, LevelsPrefixR
...
-    static TWordType _CHAR_BITMASKS[ValueSize<TValue>::VALUE]; // filter by character
-    static TWordType _TRUNC_BITMASKS[_VALUES_PER_WORD]; // truncate the last values in a word that shell not be counted
+    static TWordType _SELECT_BITMASK; // select only every 2nd element
+
+    static TWordType
+    _CHAR_BITMASKS[RankDictionaryBlockSize_<TValue, Levels<TSpec, TConfig> >::VALUE]; // filter by character
+
+    // truncate the last values in a word that shall not be counted
+    static TWordType
+    _TRUNC_BITMASKS[RankDictionaryTruncBitmaskSize_<_VALUES_PER_WORD, TValue, Levels<TSpec, TConfig> >::VALUE];
```
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: I had to insert a line break here

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 13]
```diff
@@ -893,24 +1036,37 @@ _getSuperBlockRank(RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSi
 }
 
 template <typename TSpec, typename TConfig, typename TSuperBlock, typename TPos>
-inline typename std::enable_if_t<TConfig::LEVELS == 2, typename Size<RankDictionary<bool, Levels<TSpec, TConfig> > const>::Type>
-_getSuperBlockRank(RankDictionary<bool, Levels<TSpec, TConfig> > const & dict, TSuperBlock const & superblock, TPos const pos, bool const c)
+inline
+typename std::enable_if_t<TConfig::LEVELS == 2, typename Size<RankDictionary<bool, Levels<TSpec, TConfig> > const>::Type>
```
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: But it's too long for one line :)

File: include/seqan/index/index_fm_rank_dictionary_levels.h (hunk 10 above)
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: I don't get the question

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 14]
```diff
@@ -893,24 +1036,37 @@ _getSuperBlockRank(RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSi
...
 {
     // If c == false then return the complementary rank.
     return c ? superblock : pos - _toPosInSuperBlock(dict, pos) - superblock;
 }
 
 template <typename TSpec, typename TConfig, typename TSuperBlock, typename TPos>
-inline typename std::enable_if_t<TConfig::LEVELS == 3, typename Size<RankDictionary<bool, Levels<TSpec, TConfig> > const>::Type>
-_getSuperBlockRank(RankDictionary<bool, Levels<TSpec, TConfig> > const & dict, TSuperBlock const & superblock, TPos const pos, bool const c)
+inline
+typename std::enable_if_t<TConfig::LEVELS == 3, typename Size<RankDictionary<bool, Levels<TSpec, TConfig> > const>::Type>
```
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: see above

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 15]
```diff
@@ -456,36 +520,81 @@ struct RankDictionary<TValue, Levels<TSpec, TConfig> >
     // Constructors
     // ------------------------------------------------------------------------
 
+    void _populateBitmasks()
+    {
+        // TODO(cpockrandt): make all bitmasks a const-expr
+        const unsigned maxValue = (1 << _BITS_PER_VALUE) - 1;
+        const unsigned padding = _BITS_PER_WORD % _BITS_PER_VALUE;
+        const TWordType paddingWord = (padding > 0) ? (1ull << (_BITS_PER_WORD - padding)) : 0ull;
+        if (isPRDict(*this) && !std::is_same<TValue, bool>::value)
```
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: No, it's the only function call.

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 16]
```diff
@@ -893,24 +1036,37 @@ _getSuperBlockRank(RankDictionary<TValue, Levels<TSpec, LevelsPrefixRDConfig<TSi
...
+                   bool const c)
 {
     // If c == false then return the complementary rank.
     return c ? superblock : _toPosInUltraBlock(dict, pos) - _toPosInSuperBlock(dict, pos) - superblock;
 }
 
-template <typename TSpec, typename TSize, typename TFibre, unsigned LEVELS, unsigned WPB, typename TSuperBlock, typename TPos, typename TSmaller>
-inline typename Size<RankDictionary<bool, Levels<TSpec, LevelsPrefixRDConfig<TSize, TFibre, LEVELS, WPB> > > const>::Type
-_getSuperBlockRank(RankDictionary<bool, Levels<TSpec, LevelsPrefixRDConfig<TSize, TFibre, LEVELS, WPB> > > const & dict, TSuperBlock const & superblock, TPos const pos, bool const /*c*/, TSmaller & smaller)
+template <typename TSpec, typename TSize, typename TFibre, unsigned LEVELS, unsigned WPB,
```
Comment (cpockrandt) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: Sorry, I forgot to write the macro ``TPREFIXLEVELS`` in caps :)

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 17]
```diff
@@ -345,35 +387,61 @@ struct RankDictionaryEntry_<TValue, Levels<TSpec, TConfig> >
...
+struct isLevelsPrefixRD
 {
-    return _bitmask2<TWord>(bitsTotal, blocks, blocks, blocksize, vPrefix);
+    static const bool value = false;
+};
+
+template <typename TSpec, typename TSize, typename TFibre, unsigned LEVELS, unsigned WPB>
+struct isLevelsPrefixRD<TPREFIXLEVELS>
+{
+    static const bool value = true;
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: uppercase VALUE please

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 18]
```diff
@@ -345,35 +387,61 @@ struct RankDictionaryEntry_<TValue, Levels<TSpec, TConfig> >
...
+};
+
+template <typename TDict>
+constexpr bool
+isPRDict(TDict &)
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: metafunction syntax instead

File: include/seqan/index/index_fm_rank_dictionary_levels.h (hunk 7 above)
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: re-rename agreed upon :)

File: include/seqan/index/index_fm_rank_dictionary_levels.h (hunk 10 above)
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: maybe add const to help the compiler?

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 19]
```diff
@@ -1456,9 +1727,17 @@ reserve(RankDictionary<TValue, Levels<TSpec, TConfig> > & dict, TSize const newC
     typedef RankDictionary<TValue, Levels<TSpec, TConfig> > TRankDict_;
 
     if (TConfig::LEVELS > 1)
-        reserve(dict.superblocks, (newCapacity + TRankDict_::_VALUES_PER_SUPERBLOCK - 1) / TRankDict_::_VALUES_PER_SUPERBLOCK, tag);
+    {
+        reserve(dict.superblocks,
+               (newCapacity + TRankDict_::_VALUES_PER_SUPERBLOCK - 1) / TRankDict_::_VALUES_PER_SUPERBLOCK,
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: indent one more space

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 20]
```diff
@@ -1456,9 +1727,17 @@ reserve(RankDictionary<TValue, Levels<TSpec, TConfig> > & dict, TSize const newC
...
+               tag);
+    }
     if (TConfig::LEVELS > 2)
-        reserve(dict.ultrablocks, (newCapacity + TRankDict_::_VALUES_PER_ULTRABLOCK - 1) / TRankDict_::_VALUES_PER_ULTRABLOCK, tag);
+    {
+        reserve(dict.ultrablocks,
+               (newCapacity + TRankDict_::_VALUES_PER_ULTRABLOCK - 1) / TRankDict_::_VALUES_PER_ULTRABLOCK,
```
Comment (h-2) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: here too

File: tests/index/test_index_fm_rank_dictionary.cpp [hunk 21]
```diff
@@ -78,41 +78,46 @@ typedef Levels<void, LevelsRDConfig<uint32_t, Alloc<>, 3, 3> >       Default3Lev
...
     TagList<RankDictionary<AminoAcid,       WaveletTree<> >
-    > > > > > > > > > > > > > > > > > > >
+    > > > > > > > > > > > > > > > > > > > > > > > >
     RankDictionaryPrefixSumTypes;
 
 typedef
     TagList<RankDictionary<bool,            Naive<> >,
     TagList<RankDictionary<bool,            Default1Level>,
     TagList<RankDictionary<Dna,             Default1Level>,
-    TagList<RankDictionary<Dna5Q,           Default1Level>,
```
Comment (h-2) [inline tests/index/test_index_fm_rank_dictionary.cpp]: please re-add Dna5Q

**Comment by cpockrandt** [comment]
@h-2 Should I worry that the parallel-test is failing?

**Pull request merged**